rfd/0078-login-rules.md

```diff
@@ -0,0 +1,504 @@
+
+#### `set.remove(value)`
+
+`set.remove(value)` returns a copy of the set with `value` removed.
+`set("a", "b", "c").remove("c")` returns `("a", "b")`.
+
+#### `dict(...pairs)`
+
+`dict()` returns a dictionary with keys of type `string` and values of type `set`.
+
+```
+dict(
+  pair("fruits", set("apple", "banana")),
+  pair("vegetables", set("asparagus", "brocolli")),
+)
+```
+
+returns
+
+```
+{
+  "fruits": ("apple", "banana"),
+  "vegetables": ("asparagus", "brocolli"),
+}
+```
+
+Arguments must be pairs, the first element will be the key,
+and the second element will be the value.
+
+#### `dict.add_values(key, ...values)`
+
+`dict.add_values(key, ...values)` returns a copy of the dict with the given values
+added to the set at `dict[key]`.
+If `dict[key]` is empty or it does not exist, it will be added with `values` as
+its only elements.
+
+```
+dict(
+  pair("fruits", set("apple")),
+).add_values("fruits", "banana").add("vegetables", "asparagus", "brocolli")
+```
+
+returns
+
+```
+{
+  "fruits": ("apple", "banana"),
+  "vegetables": ("asparagus", "brocolli"),
+}
+```
+
+#### `dict.remove_keys(...keys)`
+
+`dict.remove_keys(...keys)` returns a copy of the dict with the given keys
+removed.
+
+```
+dict(
+  pair("fruits", set("apple", "banana")),
+  pair("vegetables", set("asparagus", "brocolli")),
+).remove_keys("vegetables")
+```
+
+returns
+
+```
+{
+  "fruits": ("apple", "banana"),
+}
+```
+
+#### `dict.overwrite(key, values)`
+
+`dict.overwrite(key, values)` returns a copy of the dict with the given `key` set
+to `values`.
+
+```
+dict(
+  pair("fruits", set("apple", "banana")),
+  pair("vegetables", set("asparagus", "brocolli")),
+).overwrite("vegetables", set("carrot")).overwrite("trees", set("aspen"))
+```
+
+returns
+
+```
+{
+  "fruits": ("apple", "banana"),
+  "vegetables": ("carrot"),
+  "trees": ("aspen"),
+}
+```
+
+#### `ifelse(cond, value_if_true, value_if_false)`
+
+`ifelse(cond, value_if_true, value_if_false)` returns `value_if_true` if `cond`
+evaluates to `true`, else it returns `value_if_false`.
+`ifelse(set("a").contains("a"), set("b", "c"), set())`
+returns
+`("b", "c")`.
+
+#### `choose(...options)`
+
+`choose(...options)` returns the value of the first `option` for which the
+condition evaluates to `true`.
+
+`choose(option(false, set("a", "b")), option(true, set("c", "d")))`
+returns `("c", "d")`
+
+`choose(option(set("a").contains("b"), "foo"), option(set("a").contains("a"), "bar"))`
+returns `"bar"`.
+
+Use an option with the condition hardcoded to `true` to set a default value.
+
+`choose(option(set("a").contains("b"), "foo"), option(true, "default"))`
+returns `"default"`.
+
+#### `strings.replace(input, match, replacement)`
+
+Finds all matches of `match` in `input`, and replaces them with `replacement`.
+`strings.replace("user-nic", "-", "_")` returns `"user_nic"`.
+`input` can be a string or a set of strings, in which case the replacement will
+be applied to all strings in the set.
+
+#### `upper(input)`
```

Comment:
We do really want the transformation, this is meant to set the users traits, usually the comparison happens later on with allow/deny labels in roles

I think I prefer `upper` to `toupper` because it sounds more declarative than imperative